### DISCUSSION TITLE
How to deserialize to immutable collection type?

### DISCUSSION THREAD

**Discussion by Nahuel92** [micronaut-projects/micronaut-core#11386] [General] [answered]
Hi there!

I'm using Java records for my DTOs, which happen to have fields of type List/Map, and I'm trying to enforce immutability. My current DTO looks like:

```java
import com.google.common.collect.ImmutableList;

public record MyDTO(
        @JsonProperty("id") String uuid,
        @JsonProperty("items") @JsonDeserialize(as = ImmutableList.class) List<String> items){
}
```

But it throws: io.micronaut.core.beans.exceptions.IntrospectionException: No bean introspection available for type [class com.google.common.collect.ImmutableList]. Ensure the class is annotated with io.micronaut.core.annotation.Introspected.

The workaround that I could use is using compact constructors to make the collection immutable, like this:

```java
public record MyDTO(
        @JsonProperty("id") String uuid,
        @JsonProperty("items") List<String> items){
    public MyDTO {
        items = items == null ? List.of() ? items.stream()
                .filter(Objects::nonNull)
                .toList();
    }
}
```

But I wonder if there is a better approach to achieve immutability, ideally an annotation or an annotation-parameter that doesn't rely on third-party dependencies. I checked the documentation on Micronaut Serialization, and I haven't seen any annotation or configuration to enable this feature.

Any suggestion is appreciated.
Thanks!

**Comment by dstepanov** [comment]
For Micronaut Serialization, you can define custom deserializers/serializers or use `@JsonDeserialize(as` or `@SerdeConfig(deserAs`

**Reply by Nahuel92** [reply]
Hi dstepanov!
Yes, I tried with `@JsonDeserialize(as=...)` but it didn't work. I'll try with `@SerdeConfig` and will update if it works that way. Thanks!

**Comment by Nahuel92** [comment, marked as answer]
I ended up solving my issue by annotating my collection fields with either:
- `@Serdeable.Deserializable(as = ImmutableList.class)`
- `@Serdeable.Deserializable(as = ImmutableMap.class)`
